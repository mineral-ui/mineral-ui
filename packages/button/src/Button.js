/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */
import React from 'react';
import { ellipsis } from 'polished';
import {
  createStyledComponent,
  pxToEm,
  getNormalizedValue
} from '@mineral-ui/component-utils';

type Props = {
  /** Rendered content of the component */
  children?: MnrlReactNode,
  /** Displays a circular button */
  circular?: boolean,
  /** Disables the button */
  disabled?: boolean,
  /** Stretch Button to fill its container */
  fullWidth?: boolean,
  /** Icon that goes after the children*/
  iconEnd?: React$Element<*>,
  /** Icon that goes before the children */
  iconStart?: React$Element<*>,
  /** Display a minimal button */
  minimal?: boolean,
  /** Called with the click event */
  onClick?: (event: Object) => void,
  /** Display a primary button */
  primary?: boolean,
  /** Available sizes */
  size?: 'small' | 'medium' | 'large' | 'jumbo',
  /** Available types */
  type?: 'button' | 'submit',
  /** Available variants */
  variant?: 'regular' | 'danger' | 'success' | 'warning'
};

const buttonTheme = (props, baseTheme) => ({
  Button_backgroundColor: baseTheme.color_gray_20,
  Button_backgroundColor_active: baseTheme.color_gray_30,
  Button_backgroundColor_focus: baseTheme.color_gray_20,
  Button_backgroundColor_hover: baseTheme.color_gray_10,
  Button_backgroundColor_minimal_active: baseTheme.color_gray_20,
  Button_backgroundColor_minimal_hover: baseTheme.color_gray_10,
  Button_backgroundColor_primary: baseTheme.color_theme_70,
  Button_backgroundColor_primary_active: baseTheme.color_theme_80,
  Button_backgroundColor_primary_focus: baseTheme.color_theme_70,
  Button_backgroundColor_primary_hover: baseTheme.color_theme_60,
  Button_borderColor: baseTheme.borderColor,
  Button_borderColor_focus: baseTheme.color_white,
  Button_borderRadius: baseTheme.borderRadius_1,
  Button_borderWidth: 1, // px, also used to calculate padding (in ems)
  Button_boxShadow_focus: `0 0 0 1px ${baseTheme.borderColor_focus}`,
  Button_color_text: baseTheme.color_gray_100,
  Button_color_text_minimal: baseTheme.color_text_primary,
  Button_color_text_primary: baseTheme.color_text_onprimary,
  Button_fontWeight: baseTheme.fontWeight_semiBold,
  Button_padding_small: pxToEm(4),
  Button_padding_medium: pxToEm(6),
  Button_padding_large: pxToEm(8),
  Button_padding_jumbo: pxToEm(10),
  [`Button_size_${props.size}`]: baseTheme[`size_${props.size}`],

  ButtonContent_fontSize: baseTheme.fontSize_ui,
  ButtonContent_fontSize_small: pxToEm(12),
  ButtonContent_lineHeight_small: pxToEm(16),
  ButtonContent_lineHeight_medium: pxToEm(20),
  ButtonContent_lineHeight_large: pxToEm(24),
  ButtonContent_lineHeight_jumbo: pxToEm(32),
  ButtonContent_padding_small: pxToEm(3),
  ButtonContent_padding_medium: pxToEm(4),
  ButtonContent_padding_large: pxToEm(6),
  ButtonContent_padding_jumbo: pxToEm(8),

  ButtonIcon_padding_small: pxToEm(2),
  ButtonIcon_padding_medium: pxToEm(2),
  ButtonIcon_padding_large: pxToEm(4),
  ButtonIcon_padding_jumbo: pxToEm(6),

  ...baseTheme
});

const buttonStyles = (props, baseTheme) => {
  let theme = buttonTheme(props, baseTheme);

  if (props.variant !== 'regular') {
    // prettier-ignore
    theme = {
      ...theme,
      Button_backgroundColor_primary: theme[`backgroundColor_${props.variant}`],
      Button_backgroundColor_primary_active: theme[`backgroundColor_${props.variant}_active`],
      Button_backgroundColor_primary_focus: theme[`backgroundColor_${props.variant}_focus`],
      Button_backgroundColor_primary_hover: theme[`backgroundColor_${props.variant}_hover`],
      Button_boxShadow_focus: `0 0 0 1px ${theme[`borderColor_${props.variant}_focus`]}`,
      Button_color_text: theme[`color_text_${props.variant}`],
      Button_color_text_minimal: theme[`color_text_${props.variant}`],
      Button_color_text_primary: theme[`color_text_on${props.variant}`]
    };
  }

  return {
    backgroundColor: (() => {
      if (props.disabled && !props.minimal) {
        return theme.color_gray_30;
      } else if (props.primary) {
        return theme.Button_backgroundColor_primary;
      } else if (props.minimal) {
        return 'transparent';
      } else {
        return theme.Button_backgroundColor;
      }
    })(),
    borderColor:
      props.disabled || props.primary || props.minimal
        ? 'transparent'
        : theme.Button_borderColor,
    borderRadius: props.circular
      ? `${parseFloat(theme[`Button_size_${props.size}`]) / 2}em`
      : theme.Button_borderRadius,
    borderStyle: 'solid',
    borderWidth: `${theme.Button_borderWidth}px`,
    color: (() => {
      if (props.disabled) {
        return theme.color_text_disabled;
      } else if (props.primary) {
        return theme.Button_color_text_primary;
      } else if (props.minimal) {
        return theme.Button_color_text_minimal;
      } else {
        return theme.Button_color_text;
      }
    })(),
    cursor: props.disabled ? 'default' : 'pointer',
    fontWeight: theme.Button_fontWeight,
    height: theme[`Button_size_${props.size}`],
    // Because the small & medium-sized Buttons are shorter than
    // theme.fontSize_base * theme.lineHeight, the text content does not
    // vertically align correctly without setting the lineHeight equal to the
    // smallest fontSize (or smaller).
    lineHeight: theme.ButtonContent_fontSize_small,
    // Because we use boxSizing: 'border-box', we need to substract the borderWidth
    // from the padding to have the fixed height of Root and Content be correct.
    padding: `${parseFloat(theme[`Button_padding_${props.size}`]) -
      parseFloat(pxToEm(theme.Button_borderWidth))}em`,
    verticalAlign: 'middle',
    width: props.fullWidth && '100%',
    '&:focus': {
      backgroundColor: (() => {
        if (props.primary) {
          return theme.Button_backgroundColor_primary_focus;
        } else if (props.minimal) {
          return theme.Button_backgroundColor_minimal_focus;
        } else {
          return theme.Button_backgroundColor_focus;
        }
      })(),
      borderColor: theme.Button_borderColor_focus,
      boxShadow: theme.Button_boxShadow_focus
    },
    '&:hover': {
      backgroundColor: (() => {
        if (!props.disabled) {
          if (props.primary) {
            return theme.Button_backgroundColor_primary_hover;
          } else if (props.minimal) {
            return theme.Button_backgroundColor_minimal_hover;
          } else {
            return theme.Button_backgroundColor_hover;
          }
        }
      })()
    },
    // `:active` must be last, to follow LVHFA order:
    // https://developer.mozilla.org/en-US/docs/Web/CSS/:active
    '&:active': {
      backgroundColor: (() => {
        if (!props.disabled) {
          if (props.primary) {
            return theme.Button_backgroundColor_primary_active;
          } else if (props.minimal) {
            return theme.Button_backgroundColor_minimal_active;
          } else {
            return theme.Button_backgroundColor_active;
          }
        }
      })()
    },
    '&::-moz-focus-inner': { border: 0 },

    '& [role="icon"]': {
      boxSizing: 'content-box',
      fill:
        props.disabled ||
        props.primary ||
        props.minimal ||
        props.variant !== 'regular'
          ? 'currentColor'
          : theme.Button_backgroundColor_primary,
      display: 'block',
      padding: theme[`ButtonIcon_padding_${props.size}`]
    }
  };
};

const contentStyles = (props, baseTheme) => {
  const theme = buttonTheme(props, baseTheme);

  const fontSize =
    props.size === 'small'
      ? theme.ButtonContent_fontSize_small
      : theme.ButtonContent_fontSize;

  return {
    ...ellipsis('100%'),

    display: 'block',
    fontSize,
    lineHeight: getNormalizedValue(
      theme[`ButtonContent_lineHeight_${props.size}`],
      fontSize
    ),
    padding: `0 ${getNormalizedValue(
      theme[`ButtonContent_padding_${props.size}`],
      fontSize
    )}`
  };
};

const innerStyles = {
  display: 'inline-flex',
  justifyContent: 'center',
  width: '100%'
};

const Root = createStyledComponent('button', buttonStyles, {
  includeStyleReset: true
});
const Content = createStyledComponent('span', contentStyles);
const Inner = createStyledComponent('span', innerStyles);

/**
 * The Button component represents a clickable button.
 */
export default function Button({
  children,
  iconStart,
  iconEnd,
  size = 'medium',
  type = 'button',
  variant = 'regular',
  ...restProps
}: Props) {
  const rootProps = { size, type, variant, ...restProps };

  const iconSize = {
    small: 'small',
    medium: 'medium',
    large: 'medium',
    jumbo: 'large'
  };
  const startIcon =
    iconStart && React.cloneElement(iconStart, { size: iconSize[size] });
  const endIcon =
    iconEnd && React.cloneElement(iconEnd, { size: iconSize[size] });

  return (
    <Root {...rootProps}>
      <Inner>
        {startIcon}
        {children &&
          <Content size={size}>
            {children}
          </Content>}
        {endIcon}
      </Inner>
    </Root>
  );
}
