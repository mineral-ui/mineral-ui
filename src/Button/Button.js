/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */
import React, { cloneElement } from 'react';
import { ellipsis } from 'polished';
import { createStyledComponent, pxToEm, getNormalizedValue } from '../styles';

type Props = {
  /** Rendered content of the component */
  children?: React$Node,
  /** Displays a circular Button */
  circular?: boolean,
  /** Disables the Button */
  disabled?: boolean,
  /** Element to be used as the root node - e.g. `a` can be used to create a link that is styled like a Button */
  element?: $FlowFixMe, // Should allow string | React class
  /** Stretch Button to fill its container */
  fullWidth?: boolean,
  /** Icon that goes after the children*/
  iconEnd?: React$Element<*>,
  /** Icon that goes before the children */
  iconStart?: React$Element<*>,
  /** Display a minimal Button */
  minimal?: boolean,
  /** Called with the click event */
  onClick?: (event: SyntheticEvent<>) => void,
  /** Display a primary Button */
  primary?: boolean,
  /** Available sizes */
  size?: 'small' | 'medium' | 'large' | 'jumbo',
  /** Available types */
  type?: string,
  /** Available variants */
  variant?: 'regular' | 'danger' | 'success' | 'warning'
};

export const componentTheme = (baseTheme: Object) => ({
  Button_backgroundColor: baseTheme.color_gray_20,
  Button_backgroundColor_active: baseTheme.color_gray_30,
  Button_backgroundColor_focus: baseTheme.color_gray_20,
  Button_backgroundColor_hover: baseTheme.color_gray_10,
  Button_backgroundColor_minimal_active: baseTheme.color_gray_20,
  Button_backgroundColor_minimal_hover: baseTheme.color_gray_10,
  Button_backgroundColor_primary: baseTheme.color_theme_60,
  Button_backgroundColor_primary_active: baseTheme.color_theme_70,
  Button_backgroundColor_primary_focus: baseTheme.color_theme_60,
  Button_backgroundColor_primary_hover: baseTheme.color_theme_50,
  Button_borderColor: baseTheme.borderColor,
  Button_borderColor_focus: baseTheme.color_white,
  Button_borderRadius: baseTheme.borderRadius_1,
  Button_borderWidth: 1, // px
  Button_boxShadow_focus: `0 0 0 1px ${baseTheme.borderColor_focus}`,
  Button_color_text: baseTheme.color_gray_100,
  Button_color_text_minimal: baseTheme.color_text_primary,
  Button_color_text_primary: baseTheme.color_text_onprimary,
  Button_fontWeight: baseTheme.fontWeight_semiBold,
  Button_paddingHorizontal: baseTheme.space_inset_sm,
  Button_paddingIconOnly_small: pxToEm(3),
  Button_paddingIconOnly_medium: pxToEm(7),
  Button_paddingIconOnly_large: pxToEm(7),
  Button_paddingIconOnly_jumbo: pxToEm(13),
  Button_height_small: baseTheme.size_small,
  Button_height_medium: baseTheme.size_medium,
  Button_height_large: baseTheme.size_large,
  Button_height_jumbo: baseTheme.size_jumbo,

  ButtonContent_fontSize: baseTheme.fontSize_ui,
  ButtonContent_fontSize_small: pxToEm(12),

  ButtonIcon_margin: baseTheme.space_inset_sm,

  ...baseTheme
});

function chooseColor({ disabled, primary, minimal }: Props, theme) {
  if (disabled) {
    return theme.color_text_disabled;
  } else if (primary) {
    return theme.Button_color_text_primary;
  } else if (minimal) {
    return theme.Button_color_text_minimal;
  } else {
    return theme.Button_color_text;
  }
}

const styles = {
  button: props => {
    let theme = componentTheme(props.theme);
    const {
      circular,
      disabled,
      fullWidth,
      minimal,
      primary,
      size,
      text,
      variant
    } = props;

    if (variant !== 'regular') {
      // prettier-ignore
      theme = {
        ...theme,
        Button_backgroundColor_primary: theme[`backgroundColor_${variant}`],
        Button_backgroundColor_primary_active: theme[`backgroundColor_${variant}_active`],
        Button_backgroundColor_primary_focus: theme[`backgroundColor_${variant}_focus`],
        Button_backgroundColor_primary_hover: theme[`backgroundColor_${variant}_hover`],
        Button_boxShadow_focus: `0 0 0 1px ${theme[`borderColor_${variant}_focus`]}`,
        Button_color_text: theme[`color_text_${variant}`],
        Button_color_text_minimal: theme[`color_text_${variant}`],
        Button_color_text_primary: theme[`color_text_on${variant}`]
      };
    }

    const color = chooseColor(props, theme);
    return {
      backgroundColor: (() => {
        if (disabled && !minimal) {
          return theme.color_gray_30;
        } else if (primary) {
          return theme.Button_backgroundColor_primary;
        } else if (minimal) {
          return 'transparent';
        } else {
          return theme.Button_backgroundColor;
        }
      })(),
      borderColor:
        disabled || primary || minimal
          ? 'transparent'
          : theme.Button_borderColor,
      borderRadius: circular
        ? `${parseFloat(theme[`Button_height_${size}`]) / 2}em`
        : theme.Button_borderRadius,
      borderStyle: 'solid',
      borderWidth: `${theme.Button_borderWidth}px`,
      color,
      cursor: disabled ? 'default' : 'pointer',
      display: 'inline-block',
      fontWeight: theme.Button_fontWeight,
      height: theme[`Button_height_${size}`],
      // if the user puts in a small icon in a large button
      // we want to force the button to be round/square
      // (really just pertinent on icon-only buttons)
      minWidth: theme[`Button_height_${size}`],
      padding:
        text === undefined
          ? theme[`Button_paddingIconOnly_${size}`]
          : `0 ${theme.Button_paddingHorizontal}`,
      textDecoration: 'none',
      verticalAlign: 'middle',
      width: fullWidth && '100%',
      '&:focus': {
        backgroundColor: (() => {
          if (primary) {
            return theme.Button_backgroundColor_primary_focus;
          } else if (minimal) {
            return theme.Button_backgroundColor_minimal_focus;
          } else {
            return theme.Button_backgroundColor_focus;
          }
        })(),
        borderColor: theme.Button_borderColor_focus,
        boxShadow: theme.Button_boxShadow_focus,
        color,
        textDecoration: 'none'
      },
      '&:hover': {
        backgroundColor: (() => {
          if (!disabled) {
            if (primary) {
              return theme.Button_backgroundColor_primary_hover;
            } else if (minimal) {
              return theme.Button_backgroundColor_minimal_hover;
            } else {
              return theme.Button_backgroundColor_hover;
            }
          }
        })(),
        color,
        textDecoration: 'none'
      },
      // `:active` must be last, to follow LVHFA order:
      // https://developer.mozilla.org/en-US/docs/Web/CSS/:active
      '&:active': {
        backgroundColor: (() => {
          if (!disabled) {
            if (primary) {
              return theme.Button_backgroundColor_primary_active;
            } else if (minimal) {
              return theme.Button_backgroundColor_minimal_active;
            } else {
              return theme.Button_backgroundColor_active;
            }
          }
        })(),
        color
      },
      '&::-moz-focus-inner': { border: 0 },

      '& [role="icon"]': {
        boxSizing: 'content-box',
        fill:
          disabled || primary || minimal || variant !== 'regular'
            ? 'currentColor'
            : theme.Button_backgroundColor_primary,
        display: 'block',

        '&:first-child': {
          marginLeft:
            theme.direction === 'rtl' ? theme.ButtonIcon_margin : null,
          marginRight:
            theme.direction === 'ltr' ? theme.ButtonIcon_margin : null
        },

        '&:last-child': {
          marginLeft:
            theme.direction === 'ltr' ? theme.ButtonIcon_margin : null,
          marginRight:
            theme.direction === 'rtl' ? theme.ButtonIcon_margin : null
        },

        '&:only-child': {
          margin: 0
        }
      }
    };
  },
  content: props => {
    const theme = componentTheme(props.theme);
    const { size } = props;

    let paddings;

    const fontSize =
      size === 'small'
        ? theme.ButtonContent_fontSize_small
        : theme.ButtonContent_fontSize;

    if (size === undefined || size === 'large' || size === 'jumbo') {
      const padding = getNormalizedValue(
        theme.Button_paddingHorizontal,
        fontSize
      );
      paddings = {
        '&:first-child': {
          paddingLeft: theme.direction === 'ltr' ? padding : null,
          paddingRight: theme.direction === 'rtl' ? padding : null
        },

        '&:last-child': {
          paddingLeft: theme.direction === 'rtl' ? padding : null,
          paddingRight: theme.direction === 'ltr' ? padding : null
        }
      };
    }

    return {
      ...ellipsis('100%'),

      display: 'block',
      fontSize,
      lineHeight: getNormalizedValue(theme[`Button_height_${size}`], fontSize),
      ...paddings
    };
  },
  inner: {
    alignItems: 'center',
    display: 'inline-flex',
    justifyContent: 'center',
    maxHeight: '100%',
    width: '100%'
  }
};

const Content = createStyledComponent('span', styles.content);
const Inner = createStyledComponent('span', styles.inner);

function isTypeButton(type: ?string) {
  return ['button', 'submit', 'reset'].indexOf(type) !== -1;
}

function filterProps({ element, type }: Props) {
  // When element is a component, e.g. ReactRouterLink,
  // these are not filtered automatically by rootEl
  const invalidComponentProps = ['primary', 'text', 'variant'];
  const invalidLinkProps =
    element === 'a' && isTypeButton(type) ? ['type'] : [];
  return Array.prototype.concat(invalidComponentProps, invalidLinkProps);
}

/**
 * The Button component represents a clickable button.
 * Buttons draw attention to actions that can be performed in your app.
 * Buttons are used to trigger any sort of event.
 *
 * Always use the appropriate variant for your button, and write a clear, concise label to effectively communicate intention.
 */
export default function Button({
  children,
  element = 'button',
  iconStart,
  iconEnd,
  size = 'large',
  type = 'button',
  variant = 'regular',
  ...restProps
}: Props) {
  const rootProps = {
    size,
    text: children,
    type,
    variant,
    ...restProps
  };

  const Root = createStyledComponent(element, styles.button, {
    filterProps: filterProps({ element, type }),
    includeStyleReset: true,
    rootEl: element
  });

  const iconSize = {
    small: 'medium',
    medium: 'medium',
    large: pxToEm(24),
    jumbo: pxToEm(24)
  };
  const startIcon = iconStart
    ? cloneElement(iconStart, { size: iconSize[size], key: 'iconStart' })
    : null;
  const endIcon = iconEnd
    ? cloneElement(iconEnd, { size: iconSize[size], key: 'iconEnd' })
    : null;

  return (
    <Root {...rootProps}>
      <Inner>
        {startIcon}
        {children && <Content size={size}>{children}</Content>}
        {endIcon}
      </Inner>
    </Root>
  );
}
