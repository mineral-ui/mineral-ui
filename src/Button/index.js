/* @flow */
export { default } from './Button';
