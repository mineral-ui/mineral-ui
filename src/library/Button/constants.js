/* @flow */
import { pxToEm } from '../styles';

export const ICON_SIZE = {
  small: 'medium',
  medium: pxToEm(20), // TargetX Custom Style
  large: pxToEm(24),
  jumbo: pxToEm(24)
};

export const SIZE = {
  small: 'small',
  medium: 'medium',
  large: 'large',
  jumbo: 'jumbo'
};

export const VARIANT = {
  grayscale: 'grayscale', // TargetX Custom Variant

  danger: 'danger',
  success: 'success',
  warning: 'warning'
};
