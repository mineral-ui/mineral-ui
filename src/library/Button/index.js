/* @flow */
export { default } from './Button';

export * from './themes';
