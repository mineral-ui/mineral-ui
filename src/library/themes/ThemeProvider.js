/* @flow */
import React, { Children } from 'react';
import { ThemeProvider as EmotionThemeProvider } from 'emotion-theming';
import { canUseDOM } from 'exenv';
import { getDocumentFontSize, rtlAlign } from '../utils';
import EventListener from '../EventListener';
import mineralTheme from './mineralTheme';
import DocumentFontSizeContext from './DocumentFontSizeContext';
import DocumentFontSizeProvider from './DocumentFontSizeProvider';

export type Props = {
  /** Components to which the theme will be applied */
  children?: React$Node,
  /**
   * A shallow object of [theme variables](/theming#common-scenarios-theme-structure)
   * and their values or a function that provides such an object.
   */
  theme?: Object | ((baseTheme?: Object) => Object)
};
type Theme = Object | ((baseTheme?: Object) => Object);

let domLoaded = false;

const documentFontSizeCheck = (documentFontSize) => {
  if (canUseDOM) {
    const actual = parseFloat(
      window.getComputedStyle(document.documentElement).fontSize
    );
    const provided = documentFontSize
      ? getDocumentFontSize(documentFontSize)
      : 'undefined';
    if (!((actual === 16 && !documentFontSize) || actual === provided)) {
      const received =
        typeof documentFontSize === 'string'
          ? `'${documentFontSize}'`
          : documentFontSize === undefined
            ? typeof documentFontSize
            : documentFontSize;
      // prettier-ignore
      throw new Error(
        `[mineral-ui/ThemeProvider]: Expected \`theme\` to contain a \`documentFontSize\` equivalent to the document root's fontSize, '${actual}px' or '${(actual / 16) * 100}%'. Instead got: ${received}`
      );
    }
  }
};

const handleDOMContentLoaded = () => {
  domLoaded = true;
};

const Fragment = ({ children }) => children;

const manipulateTheme = (
  baseTheme?: Object = {},
  theme?: Theme = {},
  parentDocumentFontSize?: number | string
) => {
  let outTheme = typeof theme === 'function' ? theme(baseTheme) : theme;

  if (domLoaded && theme) {
    const { documentFontSize } = outTheme;
    documentFontSizeCheck(
      documentFontSize !== undefined ? documentFontSize : parentDocumentFontSize
    );
  }

  return {
    ...baseTheme,
    ...outTheme
  };
};

/**
 * ThemeProvider provides a theme to the tree of components contained within.
 * See the [theming page](/theming) for more information.
 */
const ThemeProvider = (props: Props) => (
  <DocumentFontSizeContext.Consumer>
    {(parentDocumentFontSize) => {
      const { children, theme } = props;

      const providerProps = {
        parentDocumentFontSize,
        theme
      };

      return (
        <Fragment>
          <DocumentFontSizeProvider {...providerProps}>
            <EmotionThemeProvider
              theme={(baseTheme) =>
                manipulateTheme(baseTheme, theme, parentDocumentFontSize)
              }>
              {Children.only(children)}
            </EmotionThemeProvider>
          </DocumentFontSizeProvider>
          <EventListener
            listeners={[
              {
                target: 'document',
                event: 'DOMContentLoaded',
                handler: handleDOMContentLoaded
              }
            ]}
          />
        </Fragment>
      );
    }}
  </DocumentFontSizeContext.Consumer>
);

ThemeProvider.defaultProps = {
  theme: mineralTheme
};

export default ThemeProvider;
