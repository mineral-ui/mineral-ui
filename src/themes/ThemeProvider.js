/* @flow */
import React from 'react';
import { ThemeProvider as GlamorousThemeProvider } from 'glamorous';
import mineralTheme from './mineralTheme';

type Props = {
  /** Components to which the theme will be applied */
  children?: React$Node,
  /** A shallow object of [theme variables](/theming/#common-scenarios-theme-structure) and their values */
  theme?: Object
};

/**
 * ThemeProvider provides a theme to the tree of components contained within.  See the [theming page](/theming) for more information.
 */
export default function ThemeProvider({
  children,
  theme = mineralTheme
}: Props) {
  return (
    <GlamorousThemeProvider theme={theme}>{children}</GlamorousThemeProvider>
  );
}
