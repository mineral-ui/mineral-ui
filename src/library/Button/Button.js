/* @flow */
import React, { Component, cloneElement } from 'react';
import memoizeOne from 'memoize-one';
import { createButtonRootNode, Content, Inner } from './styled';
import { ICON_SIZE, SIZE } from './constants';

import { buttonPropTypes } from './propTypes';
import type { ButtonDefaultProps, ButtonProps } from './types';

export default class Button extends Component<ButtonProps> {
  static displayName = 'Button';

  static defaultProps: ButtonDefaultProps = {
    element: 'button',
    size: SIZE.large,
    type: 'button'
  };

  static propTypes = buttonPropTypes;

  // Must be an instance method to avoid affecting other instances memoized keys
  getRootNode = memoizeOne(
    createButtonRootNode,
    (nextProps: ButtonProps, prevProps: ButtonProps) =>
      nextProps.element === prevProps.element
  );

  render() {
    const {
      children,
      disabled,
      iconStart,
      iconEnd,
      size = Button.defaultProps.size,
      type = Button.defaultProps.type,
      variant,
      ...restProps
    } = this.props;

    const rootProps = {
      disabled,
      size,
      tabIndex: disabled ? -1 : undefined,
      text: children,
      type,
      variant,
      ...restProps
    };

    const Root = this.getRootNode(this.props, Button.defaultProps);

    const startIcon = iconStart
      ? cloneElement(iconStart, { size: ICON_SIZE[size], key: 'iconStart' })
      : null;
    const endIcon = iconEnd
      ? cloneElement(iconEnd, { size: ICON_SIZE[size], key: 'iconEnd' })
      : null;

    return (
      <Root {...rootProps}>
        <Inner>
          {startIcon}
          {children && <Content size={size}>{children}</Content>}
          {endIcon}
        </Inner>
      </Root>
    );
  }
}
