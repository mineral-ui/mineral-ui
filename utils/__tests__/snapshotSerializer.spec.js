/* @flow */
import React from 'react';
import { mount } from 'enzyme';
import ThemeProvider from '../../src/themes/ThemeProvider';
import Button from '../../src/Button';

describe('snapshotSerializer', () => {
  it('Excludes verbose ThemeProvider props from snapshot', () => {
    const tree = mount(
      <ThemeProvider>
        <Button>Submit</Button>
      </ThemeProvider>
    );

    expect(tree).toMatchSnapshot();
  });
});
