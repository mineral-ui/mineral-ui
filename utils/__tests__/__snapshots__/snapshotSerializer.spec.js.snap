// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshotSerializer Excludes verbose ThemeProvider props from snapshot 1`] = `
.css-0,
[data-css-0] {
  box-sizing: border-box;
  color: #333840;
  font-family: Open Sans, -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", Roboto, Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
  font-size: 16px;
  line-height: 1.25;
  outline: 0;
  -moz-osx-font-smoothing: auto;
  -webkit-font-smoothing: antialiased;
  background-color: #ebeff5;
  border-color: #c8d1e0;
  border-radius: 0.1875em;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  display: inline-block;
  font-weight: 600;
  height: 2.5em;
  min-width: 2.5em;
  padding: 0 0.5em;
  text-decoration: none;
  vertical-align: middle;
}

.css-0 *,
[data-css-0] *,
.css-0 *::before,
[data-css-0] *::before,
.css-0 *::after,
[data-css-0] *::after {
  box-sizing: inherit;
}

.css-0:focus,
[data-css-0]:focus,
.css-0[data-simulate-focus],
[data-css-0][data-simulate-focus] {
  background-color: #ebeff5;
  border-color: #fff;
  box-shadow: 0 0 0 1px #2e6fd9;
  color: #333840;
  text-decoration: none;
}

.css-0:hover,
[data-css-0]:hover,
.css-0[data-simulate-hover],
[data-css-0][data-simulate-hover] {
  background-color: #f5f7fa;
  color: #333840;
  text-decoration: none;
}

.css-0:active,
[data-css-0]:active,
.css-0[data-simulate-active],
[data-css-0][data-simulate-active] {
  background-color: #dde3ed;
  color: #333840;
}

.css-0::-moz-focus-inner,
[data-css-0]::-moz-focus-inner,
.css-0[data-simulate-mozfocusinner],
[data-css-0][data-simulate-mozfocusinner] {
  border: 0;
}

.css-0 [role="img"],
[data-css-0] [role="img"] {
  box-sizing: content-box;
  fill: #2e6fd9;
  display: block;
}

.css-0 [role="img"]:first-child,
[data-css-0] [role="img"]:first-child {
  margin-right: 0.5em;
}

.css-0 [role="img"]:last-child,
[data-css-0] [role="img"]:last-child {
  margin-left: 0.5em;
}

.css-0 [role="img"]:only-child,
[data-css-0] [role="img"]:only-child {
  margin: 0;
}

.css-14,
[data-css-14] {
  align-items: center;
  display: -webkit-inline-box;
  display: -moz-inline-box;
  display: -ms-inline-flexbox;
  display: -webkit-inline-flex;
  display: inline-flex;
  justify-content: center;
  max-height: 100%;
  width: 100%;
  -webkit-box-align: center;
  -webkit-align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
}

.css-15,
[data-css-15] {
  display: block;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  word-wrap: normal;
  font-size: 0.875em;
  line-height: 2.857142857142857em;
}

.css-15:first-child,
[data-css-15]:first-child,
.css-15[data-simulate-firstchild],
[data-css-15][data-simulate-firstchild] {
  padding-left: 0.5714285714285714em;
}

.css-15:last-child,
[data-css-15]:last-child,
.css-15[data-simulate-lastchild],
[data-css-15][data-simulate-lastchild] {
  padding-right: 0.5714285714285714em;
}

<ThemeProvider>
  <ThemeProvider>
    <Button
      element="button"
      size="large"
      type="button"
      variant="regular"
    >
      <glamorous(button)
        element="button"
        size="large"
        text="Submit"
        type="button"
        variant="regular"
      >
        <button
          className="css-0"
          type="button"
        >
          <glamorous(span)>
            <span
              className="css-14"
            >
              <glamorous(span)
                size="large"
              >
                <span
                  className="css-15"
                >
                  Submit
                </span>
              </glamorous(span)>
            </span>
          </glamorous(span)>
        </button>
      </glamorous(button)>
    </Button>
  </ThemeProvider>
</ThemeProvider>
`;
